### PR TITLE
List to recycle

### PR DESCRIPTION
**Pull request by devilmonk**
The PR is completed and updated based on this PR: 
https://github.com/kiwix/kiwix-android/pull/1562#issue-331150552

I don't know how one updates an existing PR. I have found it out, so next time it won't happen. :)

### CODE REVIEW

**Comment by devilmonk** [comment]
@macgills Here I am done. I didn't know how to amend a PR. But now, I have found it out.

**Comment by macgills** [comment]
I think because of your username change, the previous pr says the source is `unknown repository`. Please close the non relevant PRs

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+import static javax.swing.UIManager.getString;
+import static org.kiwix.kiwixmobile.core.utils.StyleUtils.dialogStyle;
+
+public class DefaultAdapter extends RecyclerView.Adapter<DefaultAdapter.ViewHolder> {
+
+  private Context context;
+  private AdapterView.OnItemClickListener onItemClickListener;
+  List<String> searchList;
+  @Inject
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: this will not work, have you tested the code?

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import android.widget.Toast;
+import androidx.annotation.NonNull;
+import androidx.appcompat.app.AlertDialog;
+import androidx.recyclerview.widget.RecyclerView;
+import java.util.List;
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import org.kiwix.kiwixmobile.core.R;
+
+import static javax.swing.UIManager.getString;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: I hope that is unused

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 1 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: I did not get you. The butterknife?

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: `@Inject` is dagger code, you do not call inject with a component in this class so you cannot use field injection. The android team have finally published a [guide on this topic](https://developer.android.com/training/dependency-injection) I doubt this code would even compile, have you run it on a device? Is your pre-commit hook running? It would have caught this and prevented a commit

Butterknife is annotation based too but uses different ones `@BindView` primarily. Read more [here](https://jakewharton.github.io/butterknife/)

Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: Are you talking about this warning? 
```
w: warning: Supported source version 'RELEASE_6' from annotation processor 'org.jetbrains.kotlin.kapt3.base.ProcessorWrapper' less than -source '1.8'
```

I think I am a bad programmer :-(  I didn't need most of these variables.
I could get rid of OnItemClickListener, and even searchPresenter

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Nonnull
+    public ViewHolder(@NonNull View itemView) {
+      super(itemView);
+      recentSearch = itemView.findViewById(android.R.id.text1);
+      itemView.setOnClickListener(new View.OnClickListener() {
+        @Override public void onClick(View v) {
+          if (context instanceof SearchActivity) {
+            String title = recentSearch.getText().toString();
+            ((SearchActivity) context).searchPresenter.saveSearch(title);
```
Comment (devilmonk) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: @macgills This is what I was talking about. If this is not the ideal way as you said, I don't know how should I go about it.I didn't get what you said.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 1 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: Nobody starts out writing good code, quite the opposite. You just have to keep trying over and over.

Dagger is a huge topic to get into but one of its "benefits" is it has compile time safety. Maybe it won't trigger on this because a binding has been supplied for SearchPresenter but this variable _will_ be null at runtime because just annotating a field `@Inject` does nothing

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 2 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: Yes, I actually found out that many of the things: like the search presenter, OnItemClickListener, etc. were not getting used.
I think my programming skills are not neat. :-|

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 1 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: Oh! I got it. I am just getting rid of these variables.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 3 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]:
```
class DefaultAdapter{
interface ClickListener{
void onItemClick(String text);
void onItemLongClick(String text);
}
private final ClickListener clickListener;
public DefaultAdapter(ClickListener clickListener){
    this.clickListener = clickListener
  }
.
.
.
itemView.setOnClickListener(v-> clickListener.onItemClick(recentSearch.getText().toString()))
}
```

abridged but how you do it

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  @NonNull public String getItem(int position) {
+    return searchList.get(position);
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    @Nonnull
+    TextView recentSearch;
+
+    @Nonnull
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: Wrong `NonNull` annotation

**Comment by kelson42** [comment]
@devilmonk Could you please fix the last problem reported by the reviewer?

**Comment by devilmonk** [comment]
> @devilmonk Could you please fix the last problem reported by the reviewer?

Yes, give me a week (exams). I have stopped programming for a week, so on 28, I'll be back.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 3 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: can you please guide me through it. I have not done it so far in my life

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: the only really missing piece is the code to write from the calling location, here an activity but this concept of callbacks is very generalised. There are further examples [here](https://stackoverflow.com/questions/18279302/how-do-i-perform-a-java-callback-between-classes), perhaps if you try asking me very specific questions I can try and shed some light

Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: I think I have solved the issue. Please check it out.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 5]
```diff
@@ -56,18 +59,15 @@
...
-
   private final int REQ_CODE_SPEECH_INPUT = 100;
   @Inject
   SearchPresenter searchPresenter;
-  private ListView listView;
-  private ArrayAdapter<String> currentAdapter;
+  private RecyclerView recyclerView;
   private AutoCompleteAdapter autoAdapter;
-  private ArrayAdapter<String> defaultAdapter;
+  DefaultAdapter defaultAdapter;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: this should be private

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 6]
```diff
@@ -79,20 +79,19 @@ protected void onCreate(Bundle savedInstanceState) {
...
     }
+
     Toolbar toolbar = findViewById(R.id.toolbar);
     ViewCompat.setLayoutDirection(toolbar, ViewCompat.LAYOUT_DIRECTION_LOCALE);
     setSupportActionBar(toolbar);
     getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_action_back);
     getSupportActionBar().setHomeButtonEnabled(true);
     searchPresenter.attachView(this);
-    listView = findViewById(R.id.search_list);
+    recyclerView = findViewById(R.id.search_list);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: butterknife

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 7]
```diff
@@ -131,7 +128,7 @@ public void finish() {
     }
   }
 
-  @Override
+  @RequiresApi(api = Build.VERSION_CODES.M) @Override
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: I doubt this requires API M

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 8]
```diff
@@ -101,18 +100,16 @@ protected void onCreate(Bundle savedInstanceState) {
...
   public void activateAutoAdapter() {
-    currentAdapter = autoAdapter;
-    listView.setAdapter(currentAdapter);
+    recyclerView.setAdapter(autoAdapter);
   }
 
   @Override
   public void addRecentSearches(List<String> recentSearches) {
-    defaultAdapter.addAll(recentSearches);
+    defaultAdapter.searchList.addAll(recentSearches);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: don't access the fields of another class directly, SearchActivity should not know about `searchList`, make it private and add the method `addAll` to DefaultAdapter

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 9]
```diff
@@ -285,24 +266,17 @@ private void resetAdapter() {
...
-        View row;
-
-        if (convertView == null) {
-          row = LayoutInflater.from(parent.getContext())
-            .inflate(android.R.layout.simple_list_item_1, null);
-        } else {
-          row = convertView;
-        }
+  private DefaultAdapter getDefaultAdapter() {
+    return new DefaultAdapter(this, new DefaultAdapter.CLickListener() {
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Another option here is to make SearchActivity implement DefaultAdapter.CLickListener and construct DefaultAdapter like `new DefaultAdapter(this, this)`

This change isn't particularly needed but it keeps this class looking more like it used to. I'll leave it to your discrection

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java [hunk 10]
```diff
@@ -83,9 +87,17 @@ public String getItem(int index) {
...
+  @Override public int getItemCount() {
+    return data.size();
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    TextView title;
+
+    public ViewHolder(@NonNull View itemView) {
+      super(itemView);
+      title = itemView.findViewById(android.R.id.text1);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java]: butterknife

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+import android.widget.TextView;
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.RecyclerView;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+public class DefaultAdapter extends RecyclerView.Adapter<DefaultAdapter.ViewHolder> {
+
+  private Context context;
+  List<String> searchList;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: this is not initialised so it is null, I believe this crashes right now

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  @NonNull public String getItem(int position) {
+    return searchList.get(position);
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    TextView recentSearch;
+
+    public ViewHolder(@NonNull View itemView) {
+      super(itemView);
+      recentSearch = itemView.findViewById(android.R.id.text1);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: butterknife

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    this.context = context;
+    this.clickListener = clickListener;
+  }
+
+  @NonNull @Override
+  public DefaultAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+    View view = (View) LayoutInflater.from(parent.getContext())
+      .inflate(android.R.layout.simple_list_item_1, parent, false);
+    ViewHolder holder = new ViewHolder(view);
+    return holder;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: no need to create the variable, `return new ViewHolder(view)`

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import androidx.recyclerview.widget.RecyclerView;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+public class DefaultAdapter extends RecyclerView.Adapter<DefaultAdapter.ViewHolder> {
+
+  private Context context;
+  List<String> searchList;
+
+  interface CLickListener {
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: rename this (shift+f6) to `ClickListener`, I have an annoying quirk that when I type I sometimes capitalise the next letter after a capital `C`

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    void onItemClick(String text);
+    boolean onItemLongClick(String text);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: this doesn't need to return a boolean.
Boolean is the return type of View.OnLongClickListener because the view wants to know if the click event was handled or not, we don't care about that and can just always return true where we set the longClickListener, this means we can focus on the action we want to perform in our implementation

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 14 above)
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: An Interface definition should not be inbetween fields, at the top of the class is probably for the best

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+import android.view.ViewGroup;
+import android.widget.TextView;
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.RecyclerView;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+public class DefaultAdapter extends RecyclerView.Adapter<DefaultAdapter.ViewHolder> {
+
+  private Context context;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: this can be final

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 11 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]:
```
2575
01:14:39 V/InstrumentationResultParser: There was 1 failure:
2576
01:14:39 V/InstrumentationResultParser: 1) SearchActivitySimple(org.kiwix.kiwixmobile.search.SearchActivityTest)
2577
01:14:39 V/InstrumentationResultParser: Test instrumentation process crashed. Check org.kiwix.kiwixmobile.search.SearchActivityTest#SearchActivitySimple.txt for details
```
[Yup](https://travis-ci.com/kiwix/kiwix-android/builds/139140868)

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 7 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: I don't know why but if I don't do it android studio keeps on showing me error.

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: I'll clone your branch and check it out

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 15 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: Initially, I was not using boolean but that kept on showing me the error:
```
bad return type in lambda expression, void cannot be converted to boolean
```

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: that is because you are using an `expression lambda`, you will need to use a `statement lambda`
eg
```
v.setOnLongClickListener(v->{
listener.onLongClick()
return true
)
```
if you ever have trouble with lambdas you can try to write out the longform, remember that lambdas are only syntax sugar
eg
```
v.setOnLongClickListener(new View.OnLongClickListener(){
@Override public Boolean onLongClick(some arguments or something){
   listener.onLongClick()
  return true
}
})
```
and then studio will give you a prompt to convert it to a lambda

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 7 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Regardless you should never annotate a method you are not responsible for calling with this annotation
image

The reason is explained quite well by AS:
image

I assume you hit `alt+enter` and got this:
image
and then went for the first option.

What you wanted was the 2nd action but this reporting is very strange for a few reasons
1. It is only reporting the 2nd usage
1. We have 150,000 users and have never seen this crash

The underlying reason is that this is a broadcast from the android system, read more [here](https://medium.com/androiddevelopers/custom-text-selection-actions-with-action-process-text-191f792d2999), and will only happen on M and up anyways.

In summary not really a problem

**Comment by macgills** [comment]
Something I noticed because I checked out your fork:
image
Lots of unused imports.

 I personally have a macro bound to ctrl+s that does the following:
image

You may also want to check your Auto Import settings, these are mine
image

**Comment by devilmonk** [comment]
Thanks @macgills

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 7 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Oh! Got it.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 6 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Using butterknife on Toolbar is showing an error and won't compile.
```
@BindView(R.id.toolbar) Toolbar toolbar;
```
It says that: the attribute must be constant. On compiling it says that bindview can not be done on toolbar.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 11 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]:
```
  public void addAll(List<String> recentSearches) {
    searchList.addAll(recentSearches);
  }
```

Is it fine? Still, it is not being initialied but we are adding elements to it when addRecentSearches is called.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 6 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: The problem is that we are in a library module when in `core`.

Did you google `butterknife the attribute must be constant.`? It would have lead to [this issue](https://github.com/JakeWharton/butterknife/issues/468) which then links to [the solution](https://github.com/JakeWharton/butterknife#library-projects) which is the readme on github.

tl;dr use R2, instead of R

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java (hunk 11 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: You can't call methods on a null object, this is the reason for Null Pointer Exceptions.

Perhaps your confusion is arising from how `primitives` work, say int/boolean
```
int aDefaultValueInt; //same as writing int aDefaultValueInt = 0;
```

this only is true for the primitives, every other object is initialized to null

```
Integer aDefaultValueInteger; //same as writing Integer aDefaultValueInteger = null;
```
*Note the `boxed type` in the example above.

This is because when no value is specified java supplies a default value, primitives can not be null so it is `false` or `0` but the logical default value for an Object is `null`.

Initialise the field at the declaration or in the constructor so it can be marked final.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 6 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: I did but I only checked some StackOverflow's answers which did not solve the problem. I never knew that when working on core, it is a library module.

Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: There is one more
```
    Toolbar toolbar = findViewById(R.id.toolbar);
    setSupportActionBar(toolbar);
```
I think the same toolbar variable can be again used here, I feel like I can get rid of this variable. As I did not did this, do it serve any purpose? I think no.

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Use butterknife to bind any views (when using java) and if you see `findViewById` get rid of it please

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  @Override public int getItemCount() {
+    return searchList.size();
+  }
+
+  @NonNull public String getItem(int position) {
+    return searchList.get(position);
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    @BindView(android.R.id.text1) TextView recentSearch;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/DefaultAdapter.java]: this will be null, you haven't called Butterknife.bind

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java [hunk 18]
```diff
@@ -83,9 +83,16 @@ public String getItem(int index) {
...
 
-  @Override
-  public Filter getFilter() {
-    return mFilter;
+  @Override public int getItemCount() {
+    return data.size();
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    @BindView(android.R.id.text1) TextView title;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java]: This will also be null.

**Comment by macgills** [comment]
You have conflicts you must resolve, merge develop to your branch and resolve.

**Comment by macgills** [comment]
@devilmonk have you tested this by running it on a device or emulator?

**Comment by devilmonk** [comment]
> @devilmonk have you tested this by running it on a device or emulator?

No, let me do it.

**Comment by devilmonk** [comment]
@macgills it is fine, rather I will say, it is fine at least on my computer ;-)

**Comment by macgills** [comment]
Then I think it is likely you are looking at the wrong search. SearchActivity is launched when you click the search icon with a zim file open. I see no results for my search:
image

**Comment by devilmonk** [comment]
And I thought that writing software was easier than doing DataStructures/Algo questions, joking.

I don't know, my app won't even let me download any book, maybe due to slow internet connection. I mean, it keeps on loading and loading. I missed this one line in, do you think this was causing the issue?

Earlier it was:
```
  @Override public void onItemClick(String text) {
    searchPresenter.saveSearch(text);
  }
```

I have changed it to,

```
  @Override public void onItemClick(String text) {

    searchPresenter.saveSearch(text);
    sendMessage(text);
  }
```
I compared the original code on kiwix and the one I wrote on my PC, trying to make logical connections, I found this one line missing.

**Comment by macgills** [comment]
Software is quite difficult and android is particularly arse-y.

You should diagnose why you can't download a book, that is a serious impediment to development. There are a bunch of logs in logcat that should help you figure it out and failing that you have access to a debugger, an invaluable tool. If you can't download through the app you can try downloading a zim off the web from [here](http://download.kiwix.org/zim/) and the app should detect the zim file.

It is not that line. Look at what send message does.

```
  private void sendMessage(String uri) {
    int value = shouldStartNewActivity();
    if (value == 1) {
      Intent i = Intents.internal(CoreMainActivity.class);
      i.putExtra(TAG_FILE_SEARCHED, uri);
      startActivity(i);
    } else {
      Intent i = new Intent();
      i.putExtra(TAG_FILE_SEARCHED, uri);
      setResult(RESULT_OK, i);
      finish();
    }
  }
```

It either starts up MainActivity or finishes SearchActivity and returns a result to MainActivity. That line should be there but it has no bearing on why nothing is rendering because this is the action taken when you click an item in the list, in the current scenario there is nothing available to click.

You should probably review the [official guide](https://developer.android.com/guide/topics/ui/layout/recyclerview)

**Comment by kelson42** [comment]
@devilmonk ping? This PR has been open fore really too long... Any chance we could speed up and close it?

**Comment by kelson42** [comment]
No feedback, closing the PR.
